### PR TITLE
create API for searching for numbers within string - solving LANG-165…

### PR DESCRIPTION
**Pull request by Mikhail-Polivakha**
Solving Jira LANG-1658 issue. Unit test with multiple assertions is included.
Jira issue reference: https://issues.apache.org/jira/browse/LANG-1658

### CODE REVIEW

File: src/main/java/org/apache/commons/lang3/StringUtils.java [hunk 1]
```diff
@@ -9638,6 +9638,79 @@ public static String wrapIfMissing(final String str, final String wrapWith) {
...
+
+        for (int i = 0; i < stringThatContainsNumbers.length(); i++) {
+            char currentSymbol = stringThatContainsNumbers.charAt(i);
+            if (Character.isDigit(currentSymbol)) {
+                if (!hasDigitAlreadyStarted) {
+                    hasDigitAlreadyStarted = true;
+                }
+                currentNumberAsStringBuilder.append(currentSymbol);
+                continue;
+            } else if (currentSymbol == '.') {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/StringUtils.java:9683]: This will not work on locales that use a different character for the decimal marker.

File: src/main/java/org/apache/commons/lang3/StringUtils.java [hunk 2]
```diff
@@ -9638,6 +9638,79 @@ public static String wrapIfMissing(final String str, final String wrapWith) {
...
+     * sequence "90." (imagine that there is no digit right after dot) will be
+     * considered as 90.0 double.
+     *
+     * @param stringThatContainsNumbers - string, that contains number or several numbers.
+     *                                 Not necessary integers, may be numbers with float point.
+     * @return - list of numbers, that this particular string contains
+     *
+     * @throws NumberFormatException - see documentation clarification about cases when thrown above
+     */
+    public static List<Double> extractNumbersFromString(String stringThatContainsNumbers) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/StringUtils.java:9667]: The API name says Number but the return type says Double. They should match.

File: src/test/java/org/apache/commons/lang3/StringUtilsTest.java [hunk 3]
```diff
@@ -3338,6 +3341,95 @@ public void testToRootUpperCase() {
         }
     }
 
+    @Test
+    void extractNumbersFromString() {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/lang3/StringUtilsTest.java:3345]: By convention test method have the "test" prefix.

You should break up each use case in their own method IMO.

There are no tests for edge cases: null and empty string. Also a non empty string with no numbers.

File: src/test/java/org/apache/commons/lang3/StringUtilsTest.java [hunk 4]
```diff
@@ -3338,6 +3341,95 @@ public void testToRootUpperCase() {
...
+        expectedList.add(299.0);
+        Assertions.assertThrows(NumberFormatException.class, () -> StringUtils.extractNumbersFromString("737 another is 42.23.32"));
+        expectedList.clear();
+
+        expectedList.add(737.0);
+        expectedList.add(29.1);
+        expectedList.add(299.0);
+        input = "737. something29.1. 299";
+        Assertions.assertEquals(expectedList, StringUtils.extractNumbersFromString(input));
+        expectedList.clear();
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/lang3/StringUtilsTest.java:3405]: Why is this an error condition? Sentences end in a period.

How would you treat commas in a sentence when in a locale like French that uses the comma as the decimal separator?

**Review by garydgregory — CHANGES_REQUESTED** [review]
Also what about the thousands separator as in "10,000" for 10 thousand? What about the Unicode decimal separator? A true general solution is much more complex than one that focuses on a narrower case hard coded to the period decimal separator. 

The type of string parsing is too specific for Commons Lang IMO, maybe for Commons Text.

I'm curious as to what the community thinks.

File: src/main/java/org/apache/commons/lang3/StringUtils.java (hunk 1 above)
Comment (Mikhail-Polivakha) [reply src/main/java/org/apache/commons/lang3/StringUtils.java:9683]: You propose to consume separator as method parameter?

File: src/main/java/org/apache/commons/lang3/StringUtils.java (hunk 2 above)
Comment (Mikhail-Polivakha) [reply src/main/java/org/apache/commons/lang3/StringUtils.java:9667]: So, ok, and what is exactly your proposition?) Btw all the types you mentioned extends Number class, that`s why from my point of view it is valid. Explain please more to me why it does not work

File: src/test/java/org/apache/commons/lang3/StringUtilsTest.java (hunk 3 above)
Comment (Mikhail-Polivakha) [reply src/test/java/org/apache/commons/lang3/StringUtilsTest.java:3345]: Valid points, I will add some. Thanks!

File: src/main/java/org/apache/commons/lang3/StringUtils.java (hunk 1 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/lang3/StringUtils.java:9683]: You can implement an API to do that if you want but my main point is that the code should run with inputs from any locale which means handling period and commas as both the decimal and thousands separators. But, the other point I was attempting to make is that I do not feel this code belongs in Commons Lang, it feels too much like NLP code to me. It might be something for Commons Text, but it seems quite a specific use case, too much like NLP, not generic enough for a Commons library. I encourage others in the community to opine. The NLP nature makes me wonder how you would handle input like "I 'd like 10,000 nuts, 9,000 bolts, but only up to $10.5." and "Send $4.50.5 apples too please."

**Comment by garydgregory** [comment]
Closing: No reply, and this kind of text processing does not belong in Commons Lang but could be in Commons Text.
